homeassistant/components/fan/__init__.py

```diff
@@ -116,6 +125,21 @@ async def async_set_speed(self, speed: str):
         """Set the speed of the fan."""
         if speed == SPEED_OFF:
             await self.async_turn_off()
+        elif self.supported_features & SUPPORT_SET_SPEED_PERCENTAGE:
+            await self.async_set_speed_percentage(self.speed_to_percentage(speed))
+        else:
+            await self.hass.async_add_job(self.set_speed, speed)
+
+    def set_speed_percentage(self, speed: str) -> None:
+        """Set the speed of the fan, as a percentage."""
+        raise NotImplementedError()
+
+    async def async_set_speed_percentage(self, speed: int):
+        """Set the speed of the fan, as a percentage."""
+        if speed == SPEED_OFF:
+            await self.async_turn_off()
+        elif not self.supported_features & SUPPORT_SET_SPEED_PERCENTAGE:
```

Comment:
register_entity_service will guard this call from even making it to this entity since it's not supported. Also if the entity overrides this method ,it would lose this logic. So if we want this logic, we should pass a function to register_entity_service instead of a string. `async handle_set_speed_percentage(entity, data)`

Reply:
See my reply above. My intention was not to gate the availability of the 2 services, it was that if speed was supported at all both methods would be possible. If that is possible then the idea here was that an implementation would support one approach or the other. It would either implement `async_set_speed` or `async_set_speed_percentage`. It would *not* implement both. The base entity would provide a translation for the other interface.

This does seem to work. If i create a demo fan instance, I can call `async_set_speed_percentage` on it even though it doesn't implemet it. The built in mapping translates 33 to `low`, 66 to `medium`, etc. And if use the homekit_controller instance I can call `async_set_speed` on it, even though it doesn't implement it, and it will map `low` to 33, `medium` to 66, etc.